base/promotion.jl

```diff
@@ -195,7 +195,7 @@ function typejoin_union_tuple(T::Type)
         elseif U isa Union
             ci = typejoin(U.a, U.b)
         else
-            ci = U
+            ci = promote_typejoin_union(U)
```

Comment:
@vtjnash According to what you noted at https://github.com/JuliaLang/julia/pull/30485#discussion_r251042912, it's not correct to call `promote_typejoin_union` from here given that the function is marked as `@pure`. But I don't know how to fix this otherwise, as without `@pure` the result isn't inferred, which defeats the point of all this code.

Reply:
Yes, exactly as your edit says. Note that `promote_typejoin_union(U)` here is already defined to be `ci = Any`, so that is the identical change too. And we could perhaps simplify the structure a bit:
```
diff --git a/base/promotion.jl b/base/promotion.jl
index 21245f0e05..70bb4c3b53 100644
--- a/base/promotion.jl
+++ b/base/promotion.jl
@@ -168,19 +168,17 @@ function promote_typejoin_union(::Type{T}) where T
         return Any # TODO: compute more precise bounds
     elseif T isa Union
         return promote_typejoin(promote_typejoin_union(T.a), promote_typejoin_union(T.b))
-    elseif T <: Tuple
-        return typejoin_union_tuple(T)
-    else
+    elseif T isa DataType
+        T <: Tuple && return typejoin_union_tuple(T)
         return T
+    else
+        error("unreachable") # not a type??
     end
 end
 
-function typejoin_union_tuple(T::Type)
+function typejoin_union_tuple(T::DataType)
     @_pure_meta
     u = Base.unwrap_unionall(T)
-    u isa Union && return typejoin(
-            typejoin_union_tuple(Base.rewrap_unionall(u.a, T)),
-            typejoin_union_tuple(Base.rewrap_unionall(u.b, T)))
     p = (u::DataType).parameters
     lr = length(p)::Int
     if lr == 0
@@ -192,6 +190,8 @@ function typejoin_union_tuple(T::Type)
         U = Core.Compiler.unwrapva(pi)
         if U === Union{}
             ci = Union{}
+        elseif U isa UnionAll
+            return Any # TODO: compute more precise bounds
         elseif U isa Union
             ci = typejoin(U.a, U.b)
         else
```